### ISSUE TITLE
CA2000: Rule does not trigger for IAsyncDisposable objects returned from factory methods

### ISSUE THREAD

**Issue by gkurbesov** [dotnet/roslyn-analyzers#7487]
### Analyzer

**Diagnostic ID**: [CA2000](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2000): `Dispose objects before losing scope`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 8.0.307](https://dotnet.microsoft.com/en-us/download/dotnet/8.0)

### Describe the bug

The CA2000 rule currently does not trigger for objects implementing `IAsyncDisposable` when they are returned from a method, such as a factory method. This behavior is inconsistent, as the rule correctly triggers for `IAsyncDisposable` objects that are instantiated locally using a constructor. This can lead to potential resource leaks, as these objects may not be disposed of correctly.

### Steps To Reproduce

```csharp
var testFactory = new TestFactory();

// All these lines should be targeted by CA2000
var syncDisposable = testFactory.CreateSyncDisposable(); 
var syncDisposable2 = testFactory.CreateSyncDisposable2(); 
var syncDisposable3 = testFactory.CreateSyncDisposable3(); 

var complexDisposable = testFactory.CreateComplexDisposable(); 
var complexDisposable2 = testFactory.CreateComplexDisposable2();

// Method returns IAsyncDisposable and should be targeted by CA2000, but it is not
var complexDisposable3 = testFactory.CreateComplexDisposable3(); 

var complexDisposable4 = testFactory.CreateComplexDisposable4();

// All these lines should be targeted by CA2000, but they are not
var asyncDisposable = testFactory.CreateAsyncDisposable();
var asyncDisposable2 = testFactory.CreateAsyncDisposable2();
var asyncDisposable3 = testFactory.CreateAsyncDisposable3();

// But this line is targeted by CA2000
var asyncDisposableDisposable = new SomeAsyncDisposable();

public class TestFactory
{
    public ISomeSyncDisposable CreateSyncDisposable() => new SomeSyncDisposable();
    public IDisposable CreateSyncDisposable2() => new SomeSyncDisposable();
    public SomeSyncDisposable CreateSyncDisposable3() => new SomeSyncDisposable();
    public ISomeAscyncDisposable CreateAsyncDisposable() => new SomeAsyncDisposable();
    public IAsyncDisposable CreateAsyncDisposable2() => new SomeAsyncDisposable();
    public SomeAsyncDisposable CreateAsyncDisposable3() => new SomeAsyncDisposable();
    public ISomeComplexDisposable CreateComplexDisposable() => new SomeComplexDisposable();
    public IDisposable CreateComplexDisposable2() => new SomeComplexDisposable();
    public IAsyncDisposable CreateComplexDisposable3() => new SomeComplexDisposable();
    public SomeComplexDisposable CreateComplexDisposable4() => new SomeComplexDisposable();
}

public interface ISomeSyncDisposable : IDisposable { }
public interface ISomeAscyncDisposable : IAsyncDisposable { }
public interface ISomeComplexDisposable : IDisposable, IAsyncDisposable { }

public class SomeSyncDisposable : ISomeSyncDisposable
{
    public void Dispose() { }
}

public class SomeAsyncDisposable : ISomeAscyncDisposable
{
    public ValueTask DisposeAsync() => new ValueTask(Task.CompletedTask);
}

public class SomeComplexDisposable : ISomeComplexDisposable
{
    public void Dispose() { }
    public ValueTask DisposeAsync() => new ValueTask(Task.CompletedTask);
}
```

### Expected behavior

The CA2000 rule should trigger a warning when an IAsyncDisposable object is returned from a method without being properly disposed, similar to how it behaves for locally instantiated objects.

### Actual behavior

The CA2000 rule does not trigger any warnings for IAsyncDisposable objects when they are returned from methods.

## Additional context

The issue seems to occur specifically with objects implementing `IAsyncDisposable` when they are returned from methods, such as factory methods. This inconsistency with the CA2000 rule's behavior could potentially lead to overlooked resource management issues in asynchronous code.
